KB: Add a KB on programatically filtering a Grid column bound to bool? by the null values (#2270)
telerik/blazor-docs: knowledge-base/grid-state-filter-nullable-bool.md

---

> title: Programatically filter a nullable bool column in the Grid by null values

Reviewer: Use title case [capitalization](https://docs.telerik.com/style-guide/titles-and-headings#capitalization).

> title: Programatically filter a nullable bool column in the Grid by null values

Reviewer 2: You can consider a bit shortening:

```suggestion
title: Programatically Filter Nullable Bool Grid Column by Null Values
```

> description: Learn how to programatically filter nullable bool values

Reviewer: ```suggestion
description: Learn how to programmatically filter nullable bool values in the Grid component for Blazor.
```

type: how-to

> page_title: How to filter a Grid column bound to a nullable bool by the null values programatically

Reviewer: Try to use up to 50 characters in the `page_title`.

The `page_title` length should be no more than 50 characters because the total [maximum length](https://docs.telerik.com/style-guide/metadata) is 70 characters and 24 of the 70 characters are already consumed by default during build as we automatically append the ` - Telerik UI for Blazor` suffix.

Reviewer: If the title `page_title` cannot be reasonably shortened, use the same value as in the `title` filed.

slug: grid-kb-state-filter-nullable-bool
tags: grid, blazor, column, datagrid, filter, state, nullable, bool
res_type: kb
ticketid: 1658561
---

## Environment

<table>
    <tbody>
        <tr>
            <td>Product</td>

>             <td>Grid for Blazor</td>

Reviewer 2: Perhaps the KB is applicable to the TreeList too?

Author: It does

        </tr>
    </tbody>
</table>

## Description

> This KB article also answers the following questions:

Reviewer: ```suggestion
This KB article answers the following question:
```

This KB article also answers the following questions:

> - How to filter a Grid column bound to a `bool?` by the null values programatically

Reviewer: This is the same sentence as the one above (the keywords and the message are the same). You can simply delete lines 28 and 29.

Author: I did this to search by statement and by question. I am unsure if that is correct since I am not very familiar with search engines, however, I can remove one or the other.

## Solution

To filter a Grid column bound to a `bool?` by the null values programatically:

> 

Reviewer: Adding the relevant steps as a list before the code snippet will improve the KB as this will allow the developer to quickly grasp what are the cornerstones of the proposed implementation.

Using the "see how it's done in the code" approach isn't a particularly good practice.

Author: The main reason for the lack of detailed steps is that the client can trigger that in many different ways - on a button click, in the state init event handler, or any other interactive element.

The main part of the KB is to showcase that for null values our clients must use the IsNull filter operator.

Reviewer 2: OK, then mention this milestone explicitly before the code snippet, so that the reader doesn't have to deduce this on their own.

Author: Done, you can review the last commit

````CSHTML

> @using Telerik.DataSource;

Reviewer 2: A semi-colon is required only in `.cs` files.

```suggestion
@using Telerik.DataSource
```

> <TelerikButton ThemeColor="primary" OnClick="@SetGridFilter">set filtering from code</TelerikButton>

Reviewer 2: * Use `ThemeConstants` for more user-friendliness.
* Capitalize the Button text.

```suggestion
<TelerikButton ThemeColor="@ThemeConstants.Button.ThemeColor.Primary"
               OnClick="@SetGridFilter">Filter By Null</TelerikButton>
```

> To filter a Grid column bound to a `bool?` by the null values use the `IsNull` FilterOperator. You can call the `SetGridFilter` method from a `<TelerikButton>` with `OnClick` handler (as in the code snippet below), in the [OnStateInit event]({%slug grid-state%}#onstateinit) handler, or any other custom interactive element.

Reviewer: ```suggestion
To filter a Grid column bound to a `bool?` by the null values, use the `IsNull` FilterOperator. You can call the `SetGridFilter` method from a `<TelerikButton>` with an `OnClick` handler (as in the code snippet below), in the [`OnStateInit` event]({%slug grid-state%}#onstateinit) handler, or any other custom interactive element.
```

<TelerikGrid Data="@MyData" Height="400px" @ref="@GridRef"

>              Pageable="true" FilterMode="@GridFilterMode.FilterMenu">

Reviewer 2: Separate parameters on multiple lines, according to the Docs & Demos Code Style Guide.

    <GridColumns>
        <GridColumn Field="@(nameof(SampleData.Id))" Width="120px" />
        <GridColumn Field="@(nameof(SampleData.Name))" Title="Employee Name" />
        <GridColumn Field="@(nameof(SampleData.Team))" Title="Team" />
        <GridColumn Field="@(nameof(SampleData.IsOnLeave))" Title="On Leave" />
        <GridColumn Field="@(nameof(SampleData.HireDate))" Title="Hire Date" />
    </GridColumns>
</TelerikGrid>

@code {

>     private TelerikGrid<SampleData> GridRef { get; set; }

Reviewer 2: By default, new Blazor apps use `<Nullable>enable</Nullable>` and this line will generate a warning. The best practice is to make the variable nullable and check for nullability wherever the code uses `GridRef`. See https://docs.telerik.com/blazor-ui/components/grid/filter/searchbox#search-from-code

```suggestion
    private TelerikGrid<SampleData>? GridRef { get; set; }
```

    private async Task SetGridFilter()
    {

>         GridState<SampleData> desiredState = new GridState<SampleData>()

Reviewer 2: This approach will reset all other Grid state changes that the user may have made. A more real-world implementation is to take the current Grid state and make changes to it.

            {
                FilterDescriptors = new List<IFilterDescriptor>()
            {
                new CompositeFilterDescriptor()
                {
                    FilterDescriptors = new FilterDescriptorCollection()
                    {

>                         //it is important to use the IsNull filter operator when filtering by null values

Reviewer: ```suggestion
                        // Use the IsNull filter operator when filtering by null values.
```

                        new FilterDescriptor() { Member = "IsOnLeave", Operator = FilterOperator.IsNull, Value = null, MemberType = typeof(bool?) },
                    }
                }
            }

>             };

Reviewer 2: Broken tabs

        await GridRef.SetStateAsync(desiredState);
    }

>     private IEnumerable<SampleData> MyData { get; set; }

Reviewer 2: Component properties should come before methods and handlers (code style guide).

    protected override void OnInitialized()
    {

>         Random random = new Random();

Reviewer 2: Starting from .NET 6, it's recommended to use the static `Random.Shared` instead of `new Random()`.
https://andrewlock.net/building-a-thread-safe-random-implementation-for-dotnet-framework/

        MyData = Enumerable.Range(1, 30).Select(x => new SampleData
            {
                Id = x,
                Name = "name " + x,
                Team = "team " + x % 5,
                IsOnLeave = GetRandomNullableBool(x, random),
                HireDate = DateTime.Now.AddDays(-x).Date
            });
    }

    private bool? GetRandomNullableBool(int index, Random random)
    {
        if (index % 5 == 0)
        {
            return null;
        }
        return random.Next(2) == 0 ? (bool?)false : (bool?)true;
    }

    public class SampleData
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public string Team { get; set; }
        public bool? IsOnLeave { get; set; }
        public DateTime HireDate { get; set; }
    }
}
````

> ## See Also

Reviewer 2: Link to

* https://docs.telerik.com/blazor-ui/components/grid/filter/filter-menu#filter-from-code
* https://docs.telerik.com/blazor-ui/components/grid/state

type: how-to
page_title: How to Programatically Filter Nullable Bool Grid Column by Null Value
slug: grid-kb-filter-nullable-bool
tags: telerik, blazor, grid, filter, state
res_type: kb
ticketid: 1658561
---

## Environment

<table>
    <tbody>
        <tr>
            <td>Product</td>
            <td>Grid for Blazor <br /> TreeList for Blazor</td>
        </tr>
    </tbody>
</table>

## Description

This KB article explains how to programmatically filter a Grid column bound to a `bool?` by the `null` values.

## Solution

To filter a Grid column bound to a `bool?` by the `null` values:

1. Get the current [Grid state]({%slug grid-state%}).
1. Create a `CompositeFilterDescriptor` for the desired column. Add a child [`FilterDescriptor`]({%slug components/grid/filtering%}) with an `Operator` property that is equal to `FilterOperator.IsNull`.
1. Add the new `CompositeFilterDescriptor` to the `FilterDescriptors` property of the Grid state.
1. Use the Grid `SetStateAsync` method to update the Grid state.

Alternative options are to:

* Customize the Grid filter state in the [Grid `OnStateInit` event]({%slug grid-state%}#onstateinit).

> * Use a [Grid filter template]({%slug grid-templates-filter%}) to enable users to filter by `null` from the Grid's filtering UI.

Reviewer 2: ```suggestion
* [Use a Grid filter template to enable filtering by `null` from the Grid's UI]({%slug grid-kb-filterrow-bool-customization%}).
```
